KVStore: refactor and add metrics.

eloycoto opened the pull request:
This only adds a new functionality based on metrics, but the refactor the
kvstore is a bit large.

With this change `kvstore.go` now does not export any method, all the methods
are now under `ExtendedKVStore` that wraps a namespace, that enables to know
who is using the kvstore to send metrics.

Now all `BackendOperations` methods that interact with the kvstore need to have
a namespace argument, to send the metrics. The main reason to add the namespaces is
because that methods are public, so can be used by any other package. (As
example, kvstore/store/)

New metrics are the following:

```
kvstore_operations_total                            kind="read" scope="allocator" action="GetCapabilities"   1.000000
kvstore_operations_total                            action="ListPrefix" kind="read" scope="allocator"        3.000000
kvstore_operations_total                            kind="update" scope="ipcache" action="Update"            12.000000
kvstore_operations_total                            kind="update" scope="shared-store" action="Update"       28.000000
```

The reason of the labels:

- Scope, that it's the namespace, to know what package uses the kvstore.
- Kind: There are a few kind (delete, read, set) that defines what action was
  made in the kvstore integrations.
- action: This is a low level information. is not too big (12 actions) and it's
  mainly to debug in cases that the kvstore is heavily used.

The idea for the metric is:

- Have a dashboard with the kind of operation
- Have a debug dashboard that display how many interactions per scope.

This commit is related to #5382

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/5920)
<!-- Reviewable:end -->

jrajahalme commented on pkg/kvstore/etcd.go:
```diff
@@ -604,7 +609,9 @@ func (e *etcdClient) createOpPut(key string, value []byte, lease bool) *client.O
 // Update creates or updates a key
-func (e *etcdClient) Update(key string, value []byte, lease bool) error {
+func (e *etcdClient) Update(namespace, key string, value []byte, lease bool) error {
+
```
extra empty line

jrajahalme commented on pkg/kvstore/kvstore.go:
```diff
@@ -35,88 +36,128 @@ const (
+// ExtendedKVStore is a struct that export multiple kvstore methods
+type ExtendedKVStore struct {
+	namespace string
+	mutex     lock.Mutex
```
This mutex only exists to protect changes to "namespace" (via SetNamespace()), but namespace is not ever changed. Maybe make namespace immutable and remove the mutex and SetNamespace(), and make GetNamespace() lock-free?

jrajahalme commented on pkg/kvstore/kvstore.go:
```diff
@@ -35,88 +36,128 @@ const (
+}
+
+// GetKVStoreExtendedClient returns a ExtendedKVStore struct
+func GetKVStoreExtendedClient() *ExtendedKVStore {
```
As this allocated a new struct, I'd call this `NewKVStoreExtendedClient()`.

jrajahalme commented on pkg/kvstore/kvstore.go:
```diff
@@ -35,88 +36,128 @@ const (
+}
+
+// GetKVStoreExtendedClientWithNamespace returns a ExtendedKVStore struct with the given namespace.
+func GetKVStoreExtendedClientWithNamespace(namespace string) *ExtendedKVStore {
```
As this allocated a new struct, I'd call this `NewKVStoreExtendedClientWithNamespace()`.

jrajahalme replied:
Now the `namespace` here only splits metrics to multiple sets, or namespaces. On the outset, looking at the modified KV-store API, one might think that the namespace somehow has an effect on the KV-store transactions (like putting the key-value pairs into distinct namespaces.

I don't know if there is a good solution to this, as calling it `MetricsNamespace` gets too long?

joestringer replied:
2 cents, I see this kind of usage in the current PR:
```
kvClient = kvstore.GetKVStoreExtendedClientWithNamespace(subsys)
```

I think it would be just as expressive with more clarity to do something like:
```
kvClient = kvstore.NewClient().WithMetrics(subsys)
```

jrajahalme commented on pkg/kvstore/kvstore.go:
```diff
@@ -35,88 +36,128 @@ const (
+}
+
+// SetNamespace set the namespace for the given client
+func (kvClient *ExtendedKVStore) SetNamespace(namespace string) {
```
This is not used.

jrajahalme commented on pkg/kvstore/kvstore.go:
```diff
@@ -35,88 +36,128 @@ const (
+
+// GetNamespace returns the client namespace
+func (kvClient *ExtendedKVStore) GetNamespace() string {
+	kvClient.mutex.Lock()
```
No locking needed here if `namespace` is never changed.

jrajahalme commented on pkg/kvstore/kvstore.go:
```diff
@@ -35,88 +36,128 @@ const (
-func Get(key string) ([]byte, error) {
-	v, err := Client().Get(key)
+// ExtendedKVStore is a struct that export multiple kvstore methods
+type ExtendedKVStore struct {
```
I'd rather call this just `KVStore`. The names of functions in this PR get too long otherwise.

joestringer replied:
Isn't it a `Client`? Or if you want to keep distinct names, a `KVClient` / `KVStoreClient`?

jrajahalme reviewed (changes requested):
I think that applying the `namespace` concept to the KV-store API needs clarification, and maybe discussion with people more involved with the KV-store interfacing than me. On the outset, it could be confusing that the `namespace` has no effect on the KV-store transactions (but for metrics only).

There is also some unnecessary locking here.

joestringer reviewed (commented):
I also found the "namespace" usage confusing in `pkg/kvstore/backend.go`. It's not immediately obvious to me why each of the kvstore implementations (consul/etcd) needed to be updated to take this new parameter and increment the metrics. Can we just convert all KVStore users over to use the new `ExtendedKVStore` and put the metrics counting in those methods?

eloycoto committed:
KVStore: Add metrics.

Added a new metric related to the KVstore, the scope will be the key
prefix until v[0-9]* and the dashboard example will be the following:

- Have a dashboard with the kind of operation (set/read/delete)
- Have a debug dashboard that display how many interactions per scope
(prefix)

The example metrics are:
````
kvstore_operations_total                            action="CreateOnly" kind="set" scope="cilium/state/identities/v1"    216.000000
kvstore_operations_total                            action="ListPrefix" kind="read" scope="cilium/state/identities/v1"   18.000000
kvstore_operations_total                            action="Update" kind="set" scope="cilium/state/identities/v1"        4.000000
kvstore_operations_total                            action="Update" kind="set" scope="cilium/state/ip/v1"                110.000000
kvstore_operations_total                            kind="set" scope="cilium/state/nodes/v1" action="Update"             57.000000
```

This commit is related to #5382

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

tgraf commented on pkg/kvstore/etcd.go:
```diff
@@ -22,7 +22,6 @@ import (
 
 	"github.com/cilium/cilium/pkg/controller"
 	"github.com/cilium/cilium/pkg/lock"
-
```
The removal here is likely unwanted

tgraf commented on pkg/metrics/metrics.go:
```diff
@@ -26,7 +26,6 @@ import (
 	"os"
 
 	"github.com/cilium/cilium/api/v1/models"
-
```
Unwanted

tgraf commented on pkg/kvstore/metrics.go:
```diff
@@ -0,0 +1,48 @@
+)
+
+var (
+	re = regexp.MustCompile(prefixKey)
```
I think we can simplify this and just ignore everything after the 4th slash. If there are not 4 slashes, include the first 10 characters to avoid messing up. Something like this:
```
s := strings.SplitN(key, "/", 4)
if len(s) != 4 {
    return key[:12]
} else {
    return s[3]
}
```

eloycoto replied:
Umm, your proposal does not work in case of identities: 
https://play.golang.org/p/Hp1mrt61G_u

I wouild say something like this: 

```
	s := strings.SplitN(key, "/", 5)
	if len(s) != 5 {
		return key[:12]
	} else {
		return fmt.Sprintf("%s/%s", s[2], s[3])
	}
```

tgraf reviewed (changes requested):
Looks great! Only one change request to not use regexp as I'm not sure we can rely on a number being in the key.
